src/arkode/arkode_arkstep.c

```diff
@@ -2104,6 +2160,201 @@ int arkStep_TakeStep_Z(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+  adjoint of the ERK method.
+
+  Since we are not doing error control during the adjoint integration,
+  the output variable dsmPtr should should be 0.
+
+  The input/output variable nflagPtr is used to gauge convergence
+  of any algebraic solvers within the step. In this case, it should
+  always be 0 since we do not do any algebraic solves.
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int arkStep_TakeStep_ERK_Adjoint(ARKodeMem ark_mem, sunrealtype* dsmPtr,
+                                 int* nflagPtr)
+{
+  int retval = ARK_SUCCESS;
+
+  ARKodeARKStepMem step_mem;
+
+  /* access ARKodeARKStepMem structure */
+  retval = arkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+#if SUNDIALS_LOGGING_LEVEL >= SUNDIALS_LOGGING_DEBUG
+  SUNLogger_QueueMsg(ARK_LOGGER, SUN_LOGLEVEL_DEBUG,
+                     "ARKODE::arkStep_TakeStep_ERK_Adjoint", "start-step",
+                     "step = %li, h = %" RSYM ", dsm = %" RSYM ", nflag = %d",
+                     ark_mem->nst, ark_mem->h, *dsmPtr, *nflagPtr);
+#endif
+  /* local shortcuts for readability */
+  SUNAdjointStepper adj_stepper = (SUNAdjointStepper)ark_mem->user_data;
+  sunrealtype* cvals            = step_mem->cvals;
+  N_Vector* Xvecs               = step_mem->Xvecs;
+  N_Vector sens_np1             = ark_mem->yn;
+  N_Vector sens_n               = ark_mem->ycur;
+  N_Vector sens_tmp             = step_mem->sdata;
+  N_Vector Lambda_tmp           = N_VGetSubvector_ManyVector(sens_tmp, 0);
+  N_Vector lambda_np1           = N_VGetSubvector_ManyVector(sens_np1, 0);
+  N_Vector* stage_values        = step_mem->Fe;
+
+  /* determine if method has fsal property */
+  sunbooleantype fsal = (SUNRabs(step_mem->Be->A[0][0]) == ZERO) &&
+                        ARKodeButcherTable_IsStifflyAccurate(step_mem->Be);
+
+  /* Loop over stages */
+  for (int is = step_mem->stages - 1; is >= 0; --is)
+  {
+    if (fsal && is == step_mem->stages - 1)
+    {
+      N_VConst(SUN_RCONST(0.0), stage_values[is]);
+      continue;
+    }
+
+    /* which stage is being processed -- needed for loading checkpoints */
+    ark_mem->adj_stage_idx = is;
+
+    /* Set current stage time(s) and index */
+    ark_mem->tcur = ark_mem->tn +
+                    ark_mem->h * (SUN_RCONST(1.0) - step_mem->Be->c[is]);
+
+    /*
+     * Compute partial current stage value \Lambda
+     */
+    int nvec = 0;
+    for (int js = is + 1; js < step_mem->stages; ++js)
+    {
+      /* h sum_{j=i}^{s} A_{ji}/b_i \Lambda_{j} */
+      if (step_mem->Be->b[is] > SUN_UNIT_ROUNDOFF)
+      {
+        cvals[nvec] = -ark_mem->h * step_mem->Be->A[js][is] / step_mem->Be->b[is];
```

Comment:
This is using the form where it factors out the $b^E_i$ which requires the check if it's 0 and the division here. The form that's written in the docs seems simpler. Could that form be used here?
https://sundials--559.org.readthedocs.build/en/559/arkode/Mathematics_link.html#equation-arkode-erk-adjoint